velox/functions/prestosql/tests/JsonFunctionsTest.cpp

```diff
@@ -54,6 +96,169 @@ TEST_F(JsonFunctionsTest, isJsonScalar) {
+       R"({"k1":"v1"})"_sv,
+       R"({"k1":[0,1,2]})"_sv,
+       R"({"k1":""})"_sv},
+      {"$.k1"_sv, "$.k2"_sv, "$.k1.k3"_sv, "$.k1"_sv, "$.k1"_sv},
+      {"v1"_sv, std::nullopt, std::nullopt, std::nullopt, ""_sv});
+
+  // Nested
+  evaluateWithJsonType(
+      {R"({"k1":{"k2": 999}})"_sv,
+       R"({"k1":[1,2,3]})"_sv,
+       R"({"k1":[1,2,3]})"_sv,
+       R"([{"k1":"v1"}, 2])"_sv,
+       R"([{"k1":"v1"}, 2])"_sv,
+       R"([{"k1":[{"k2": ["v1", "v2"]}]}])"_sv},
+      {"$.k1.k2"_sv,
+       "$.k1[0]"_sv,
+       "$.k1[2]"_sv,
+       "$[0].k1"_sv,
+       "$[1]"_sv,
+       "$[0].k1[0].k2[1]"_sv},
+      {"999"_sv, "1"_sv, "3"_sv, "v1"_sv, "2"_sv, "v2"_sv});
+}
+
+TEST_F(JsonFunctionsTest, jsonExtractScalarUtf8) {
+  EXPECT_EQ(
+      json_extract_scalar(R"({"k1":"I \u2665 UTF-8"})", "$.k1"),
+      "I \u2665 UTF-8");
+  EXPECT_EQ(
+      json_extract_scalar("{\"k1\":\"I \u2665 UTF-8\"}", "$.k1"),
+      "I \u2665 UTF-8");
+
+  EXPECT_EQ(
+      json_extract_scalar(
+          "{\"k1\":\"I \U0001D11E playing in G-clef\"}", "$.k1"),
+      "I \U0001D11E playing in G-clef");
+}
+
+TEST_F(JsonFunctionsTest, jsonExtractScalarInvalidPath) {
+  EXPECT_THROW(json_extract_scalar(R"([0,1,2])", ""), VeloxUserError);
+  EXPECT_THROW(json_extract_scalar(R"([0,1,2])", "$[]"), VeloxUserError);
+  EXPECT_THROW(json_extract_scalar(R"([0,1,2])", "$[-1]"), VeloxUserError);
+  EXPECT_THROW(json_extract_scalar(R"({"k1":"v1"})", "$k1"), VeloxUserError);
+  EXPECT_THROW(json_extract_scalar(R"({"k1":"v1"})", "$.k1."), VeloxUserError);
+  EXPECT_THROW(json_extract_scalar(R"({"k1":"v1"})", "$.k1]"), VeloxUserError);
+  EXPECT_THROW(json_extract_scalar(R"({"k1":"v1)", "$.k1]"), VeloxUserError);
+}
+
+// TODO: Folly tries to convert scalar integers, and in case they are large
+// enough it overflows and throws conversion error. In this case, we do out best
+// and return NULL, but in Presto java the large integer is returned as-is as a
+// string.
+TEST_F(JsonFunctionsTest, jsonExtractScalarOverflow) {
+  EXPECT_EQ(
+      json_extract_scalar(
+          R"(184467440737095516151844674407370955161518446744073709551615)",
+          "$"),
+      std::nullopt);
+}
+
+TEST_F(JsonFunctionsTest, jsonExtractSimple) {
+  EXPECT_EQ(json_extract(R"(1)", "$"), R"(1)");
+  EXPECT_EQ(json_extract(R"(123456)", "$"), R"(123456)");
+  EXPECT_EQ(json_extract(R"("hello")", "$"), R"("hello")");
+  EXPECT_EQ(json_extract(R"(1.1)", "$"), R"(1.1)");
+  EXPECT_EQ(json_extract(R"("")", "$"), R"("")");
+  EXPECT_EQ(json_extract(R"(true)", "$"), R"(true)");
+  EXPECT_EQ(json_extract(R"({"k1":"v1"})", "$.k1"), R"("v1")");
+  EXPECT_EQ(json_extract(R"([1,2])", "$[0]"), R"(1)");
+  EXPECT_EQ(json_extract(R"({"k1":{"k2": 999}})", "$.k1.k2"), R"(999)");
+  EXPECT_EQ(json_extract(R"({"k1":[1,2,3]})", "$.k1[0]"), R"(1)");
+  EXPECT_EQ(json_extract(R"([{"k1":"v1"}, 2])", "$[0].k1"), R"("v1")");
+  EXPECT_EQ(
+      json_extract(R"([{"k1":[{"k2": ["v1", "v2"]}]}])", "$[0].k1[0].k2[1]"),
```

Comment:
Please add one more element here so that $[0] is demonstrated to pick the correct one. Also try one that picks $[1].